cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -1604,19 +1623,145 @@ String zkConnectionString(int connectToReplicas, Function<Integer, String> zkNod
+                }
+                Secret coKeySecret = compositeFuture.resultAt(1);
+                if (coKeySecret == null) {
+                    return Future.failedFuture(Util.missingSecretException(namespace, ClusterOperator.secretName(cluster)));
+                }
+
+                Promise<Admin> blockingPromise = Promise.promise();
+                vertx.createSharedWorkerExecutor("kubernetes-ops-pool").executeBlocking(
+                    future -> {
+                        try {
+                            future.complete(adminClientProvider.createAdminClient(hostname, clusterCaCertSecret, coKeySecret, "cluster-operator"));
+                        } catch (RuntimeException e) {
+                            future.fail(e);
+                        }
+                    }, blockingPromise);
+                return blockingPromise.future();
+            });
+        }
+
+        /**
+         * If the kafka pod is ready, change its configuration dynamically if both necessary and possible.
+         * If it is not possible set the flag for specific podID an it will be rolled in {@code kafkaRollingUpdate}.
+         * @return Succeeded future
+         */
+        Future<ReconciliationState> kafkaBrokerDynamicConfiguration() {
+            // patches scale up/down situations
+            int runningBeforeAndAfterReplicas = Math.min(oldKafkaReplicas, kafkaCluster.getReplicas());
+            List<Future> configFutures = new ArrayList<>(runningBeforeAndAfterReplicas);
+
+            for (int podId = 0; podId < runningBeforeAndAfterReplicas; podId++) {
+                int finalPodId = podId;
+                configFutures.add(podOperations.readiness(namespace, KafkaResources.kafkaPodName(name, finalPodId), 1_000, operationTimeoutMs)
+                    .compose(ignore ->
+                            adminClient(adminClientProvider, namespace, name, finalPodId)
+                    .compose(ac -> {
+                        if (ac == null) {
+                            kafkaPodsUpdatedDynamically.put(finalPodId, null);
+                            log.warn("{}: Could not create admin client for pod {}.", reconciliation, finalPodId);
+                            return Future.succeededFuture();
+                        } else {
+                            log.debug("{}: Determining whether kafka broker {} can be update dynamically", reconciliation, finalPodId);
+                            Promise<Void> p = Promise.promise();
+                            brokerConfig(ac, finalPodId)
+                                    .compose(brokerConfig -> {
+                                        log.trace("{}: Broker description {}", reconciliation, brokerConfig);
+                                        KafkaBrokerConfigurationDiff configurationDiff = new KafkaBrokerConfigurationDiff(brokerConfig, this.kafkaConfig, kafkaCluster.getKafkaVersion(), finalPodId);
```

Comment:
`this.kafkaConfig` is quite a large String containing the Kafka properties which got built by `KafkaBrokerConfigurationBuilder` and will be used to update the ancillary CM. By retaining a reference to it we're holding it in memory for a long time (so it will likely survive a few GCs and get promoted to a different JVM GC generation, meaning a major GC will be needed to eventually collect it). It's also duplicating quite a lot of the data which is also held in the `KafkaCluster` (i.e. wasting memory). Finally, because it's a String you've had to go to the effort of parsing it in `KafkaBrokerConfigurationDiff` in order to have structured data to compute the diff.

All in all, I think this could be improved by:

* No caching the String on the ReconciliationState, but computing it fresh each time it's needed (which can easily be done in this PR)
* Allowing the `KafkaBrokerConfigurationBuilder` to be able to build a `Map` or other structured representation of the config, avoiding the need to parse it (which is a bigger refactoring which can be done in another PR).